components/sequencer/mongo/mongo_sequencer.go

```diff
@@ -0,0 +1,233 @@
+//
+// Copyright 2021 Layotto Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package mongo
+
+import (
+	"context"
+	"fmt"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+	"go.mongodb.org/mongo-driver/mongo/readconcern"
+	"go.mongodb.org/mongo-driver/mongo/writeconcern"
+	"mosn.io/layotto/components/pkg/utils"
+	"mosn.io/layotto/components/sequencer"
+	"mosn.io/pkg/log"
+)
+
+type MongoSequencer struct {
+	factory utils.MongoFactory
+
+	client     utils.MongoClient
+	session    utils.MongoSession
+	collection utils.MongoCollection
+	singResult utils.MongoSingleResult
+	metadata   utils.MongoMetadata
+	biggerThan map[string]int64
+
+	logger log.ErrorLogger
+
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+type SequencerDocument struct {
+	Id              string `bson:"_id"`
+	Sequencer_value int64  `bson:"sequencer_value"`
+}
+
+// MongoSequencer returns a new mongo sequencer
+func NewMongoSequencer(logger log.ErrorLogger) *MongoSequencer {
+	m := &MongoSequencer{
+		logger: logger,
+	}
+
+	return m
+}
+
+func (e *MongoSequencer) Init(config sequencer.Configuration) error {
+	var document SequencerDocument
```

Comment:
I found this `Init` func is duplicated with the other `Init` func in `mongo_lock.go` which is also written by you, maybe it will be better to refactor this?

Reply:
You can extract the duplicated code into a new file and reuse it. For example , code in `components/pkg/utils/etcd.go` is used by both etcd lock component and etcd sequencer component .

Reply:
ok,I will refactor the code for lock and Sequencer